### PR TITLE
Add one more pass to dump the quantized onnx model in Sophon BM188x backend

### PR DESCRIPTION
**Pull request by JerryShih**

### CODE REVIEW

**Comment by JerryShih** [comment]
It will be checked at https://github.com/ONNC/onnc/blob/630eabf47f3ec214675f34db22b6fe276b238c60/lib/Core/PassManager.cpp#L67

The second ONNXDumpOpt pass is used to dump the quantized model. So, I need to register the ONNXDumpOpt twice.

**Comment by lubatang** [comment]
PassManager has two important data structures: `execution list` and `dependency graph`. The `execution list` is a list of user-specific execution order of passes. All passes listed in execution list shall be executed no matter it is added more than once. The `dependency graph` is used to describe the dependencies among passes. It represents the static relationships between passes. Therefore, we avoid from add the same pass into the graph twice and thus you get code after  Line 67 in 630eabf.

PassManager::run() will traverse the `execution list` and to see whether there is any dependent passes or not needed to run when the module was revised. PassManager::step() is the real core of the running algorithm.

**Comment by lubatang** [comment]
https://github.com/ONNC/onnc/blob/master/tools/unittests/PassManagerTest.cpp#L235

unittest run_test1 illustrates how to add the same pass twice in the execution list. Afraid we get another issue.

**Comment by lubatang** [comment]
So? I still not see the demand to allow arbitrary ID for a pass. ONNC disallows pass to change its ID (and I believe no compiler framework allow it). I didn't saw these two instances having different dependencies to the other passes and these two instances should have identical ID.

Please found another way to implement it.

**Comment by lubatang** [comment]
Here is a document about how to write an LLVM pass. ONNC pass interface is very similar to LLVM pass. We borrow the design concept from LLVM. For your reference.
http://llvm.org/docs/WritingAnLLVMPass.html

**Comment by JerryShih** [comment]
If the two pass object a1 and a2 have the same id, the passManager is just add one more execution for the a1. Then, we will have two a1 pass.
Is it clear?

void PassManager::doAdd(Pass* pPass, TargetBackend* pBackend, State& pState)
{
  pState.execution.push_back(pPass->getPassID());

  // If the pass is already in the dependency graph, then we don't
  // need to add it into the graph.
  if (hasAdded(pPass->getPassID()))
    return;

**Comment by lubatang** [comment]
yes. two a1 class pass. But they might not the same instance.

File: lib/Target/Sophon/BM188x/ONNXDumpOptPass.cpp [hunk 1]
```diff
@@ -22,22 +22,18 @@ namespace {
...
-  static char ID;
-
 public:
   ONNXDumpOpt(BM1880Backend *pBackend) : ModulePass(ID), m_pBackend(pBackend) {}
 
   Pass::ReturnType runOnModule(Module &pModule) override;
 
 private:
   BM1880Backend *m_pBackend; // NOLINT
+  char mID;
```
Comment (lubatang) [inline lib/Target/Sophon/BM188x/ONNXDumpOptPass.cpp]: ONNC disallows arbitrary ID of a pass class.

**Comment by JerryShih** [comment]
And I need a1 and a2 instead of two a1 passes.

**Comment by lubatang** [comment]
Pls follow Liskov Substitution principle

**Pull request merged**